File: eggholder_problem/eggholderEA.py
#!/usr/bin/env python3

""" eggholderEA.py: Implements an evolutionary algorithm for the eggholder
problem
"""

__author__ = 'Georgios Kouros'
__license__ = 'BSDv3'

import numpy as np
from matplotlib import pyplot as plt
from mpl_toolkits.mplot3d import Axes3D


def egg_holder_ea() -> None:
    """ main function of the evolutionary algorithm solving the eggholder
    problem
    """
    alpha = 0.05  # mutation probability
    lambda_ = 100  # population and offspring size
    k = 3  # tournament selection
    int_max = 500  # Boundary of the domain, not intended to be changed

    # initialize population
    population = int_max * np.random.rand(lambda_, 2)

    plot_population(population, int_max)
    return

    for idx in range(0, 20):
        selected = selection(population, k)
        offspring = crossover(selected)
        joined_population = np.vstack((mutation(offspring, alpha, int_max),
                                       population))
        population = elimination(joined_population, lambda_)

        # show progress
        print(f'Iteration: {idx}, Mean fitness: {np.mean(objf(population))}')

        plot_population(population, int_max)


def objf(x: np.array) -> np.array:
    """ Computes the objective fucnction at the vector of (x,y) values

    Args:
        x (np.array): population

    Returns:
        np.array: 1D array containing the cost of the population samples
    """
    x = x.reshape((-1, 2))
    sas = np.sqrt(np.abs(x[:, 0] + x[:, 1]))
    sad = np.sqrt(np.abs(x[:, 0] - x[:, 1]))
    f_x = -x[:, 1] * np.sin(sas) - x[:, 0] * np.sin(sad)

    return f_x


def plot_population(population: np.array, int_max: int) -> None:
    """ Plot the population

    Args:
        population (np.array): The population to plot
        int_max (int): Boundary
    """
    x = np.linspace(0, int_max, 500).reshape((-1, 1))
    y = np.linspace(0, int_max, 500).reshape((-1, 1))
    F = -y.T * np.sin(np.sqrt(np.abs(x + y.T))) - x * np.sin(np.sqrt(np.abs(x - y.T)))

    Y, X = np.meshgrid(x, y)


    fig = plt.figure(1)
    ax = fig.gca(projection='3d')
    ax.plot_surface(X, Y, F,
                    rstride=8, cstride=8, shade=False, cmap="jet", linewidth=1, alpha=0.5)
    ax.scatter(population[:, 0],
               population[:, 1],
               objf(population)+1e-1, color='red')
    ax.set_xlabel('X')
    ax.set_ylabel('Y')
    ax.set_zlabel('Objective Funtion')
    ax.view_init(elev=80, azim=-135)

    plt.show()


def selection(population: np.array, k: int) -> np.array:
    """ Perform k-tournament selection to select pairs of parents """
    selected = zeros(2 * len(population), 2)

    for idx in range(0, 2 * lambda_):
        ri = randperm(lambda_, k)  # TODO replace
        _, mi = objf(population[ri, :]).min()
        selected[idx, :] = population[ri[mi], :]

    return selected


def crossover(selected: np.array) -> np.array:
    """ Perform crossover as in the slides

    Args:
        selected (np.array): The parents to produce the offsprings from

    Returns:
        np.array: The offsprings produced from the crossover operation
    """
    weights = 3 * np.random.rand(lambda_, 2) - 1
    offspring = np.zeros(lambda_, 2)

    for idx in range(1, len(offspring)):
        offspring[idx, 0] = min(int_max, max(0, selected[2 * idx - 1, 0]) + \
                weights[idx, 0] *\
                (selected[2 * idx, 0] - selected[2 * idx - 1, 0]))
        offspring[idx, 1] = min(int_max, max(0, selected[2 * idx - 1, 1]) + \
                weights[idx, 1] *\
                (selected[2 * idx, 1] - selected[2 * idx - 1, 1]))

    return offspring


def mutation(offspring: np.array, alpha: float, int_max: int) -> np.array:
    """ Perform mutation. adding a random gaussian perturbation

    Args:
        offspring (np.array): The produced offspring of the latest iteration
        alpha (float): Offset of offspring selection

    Returns:
        np.array: Mutated offspring
    """
    indices = np.random.rand(len(offspring), 1) <= alpha
    offspring[indices, :] = offspring[indices, :] + \
        10 * np.random.rand(len(indices, 2))
    offspring[indices, 0] = min(int_max, max(0, offspring[indices, 0]))
    offspring[indices, 1] = min(int_max, max(0, offspring[indices, 1]))

    return mutation


def elimination(joined_population: np.array, keep: int) -> np.array:
    """ Eliminate the unfit candidate solutions

    Args:
        joined_population (np.array): The population with the new offspring
        keep (int): The number of samples to keep from the population

    Returns:
        np.array: The samples that survived the elimination process
    """
    fvals = objf(joined_population)
    _, perm = fvals.sort()
    survivors = joined_population[perm[1:keep], :]

    return survivors


if __name__ == '__main__':
    print('Running the eggholder evolutionary algorithmic solver')
    egg_holder_ea()

File: eggholder_problem/eggholderEA.m
function eggholderEA()
  alpha = 0.05;     % Mutation probability
  lambda = 100;     % Population and offspring size
  k = 3;            % Tournament selection
  intMax = 500;     % Boundary of the domain, not intended to be changed.

  %% Initialize population
  population = intMax * rand(lambda, 2);

  plotPopulation(population, 0)

  return
  for i = 1 : 20
    %% The evolutionary algorithm
    selected = selection(population, k);
    offspring = crossover(selected);
    joinedPopulation = [mutation(offspring, alpha); population];
    population = elimination(joinedPopulation, lambda);

    %% Show progress
    display(sprintf('Mean fitness: %e', mean(objf(population))))

    plotPopulation(population, i)
  end

% Compute the objective function at the vector of (x,y) values.
function [f] = objf(x)
  sas = sqrt(abs(x(:,1)+x(:,2)));
  sad = sqrt(abs(x(:,1)-x(:,2)));
  f = -x(:,2).*sin(sas) - x(:,1).*sin(sad);
end

% Plot the population.
function plotPopulation(population, i)
  x = linspace(0,intMax,500)';
  y = linspace(0,intMax,500)';
  F = -y' .* sin(sqrt(abs(x + y'))) -x .* sin(sqrt(abs(x - y')));

  surfc(x, y, F', 'FaceLighting', 'gouraud', 'EdgeColor', 'none', 'FaceAlpha', 0.95)
  hold on
  for j = 1 : lambda
     plot3(population(j,1), population(j,2), objf(population(j,:))+1e-1, 'rx', 'LineWidth', 2)
     campos([    -3.651359848873533e+02    -4.073576307439369e+02     1.599738318114707e+04]);
     hold all;
  end
  hold off;
  drawnow;

  % Enable this if you want to print the plots to a file
%     title(sprintf('Mean fitness: %e', mean(objf(population))))
%     set(gcf,'color','w');
%     frame = getframe(1);
%     im = frame2im(frame);
%     [imind,cm] = rgb2ind(im,256);
%     imwrite(imind,cm,sprintf('EvolutionaryAlgorithmZZZ%d.png', i),'png');
%     pause;
  % Run the following command in the terminal to create gifs:
  % ffmpeg -framerate 5 -i EvolutionaryAlgorithm%d.png output.gif
end

% Perform k-tournament selection to select pairs of parents.
function [selected] = selection(population, k)
  selected = zeros(2*lambda, 2);
  for ii = 1 : 2*lambda
    ri = randperm(lambda, k);
    [~, mi] = min( objf(population(ri, :)) );
    selected(ii,:) = population(ri(mi),:);
  end
end

% Perform crossover as in the slides.
function [offspring] = crossover(selected)
  weights = 3*rand(lambda,2) - 1;
  offspring = zeros(lambda, 2);
  for ii = 1 : size(offspring,1)
    offspring(ii,1) = min(intMax,max(0,selected(2*ii-1, 1) + weights(ii,1)*(selected(2*ii, 1)-selected(2*ii-1, 1))));
    offspring(ii,2) = min(intMax,max(0,selected(2*ii-1, 2) + weights(ii,2)*(selected(2*ii, 2)-selected(2*ii-1, 2))));
  end
end

% Perform mutation, adding a random Gaussian perturbation.
function [offspring] = mutation(offspring, alpha)
  ii = find(rand(size(offspring,1),1) <= alpha);
  offspring(ii,:) = offspring(ii,:) + 10*randn(length(ii),2);
  offspring(ii,1) = min(intMax, max(0, offspring(ii,1)));
  offspring(ii,2) = min(intMax, max(0, offspring(ii,2)));
end

% Eliminate the unfit candidate solutions.
function [survivors] = elimination(joinedPopulation, keep)
  fvals = objf(joinedPopulation);
  [~, perm] = sort(fvals);
  survivors = joinedPopulation(perm(1:keep),:);
end

end
